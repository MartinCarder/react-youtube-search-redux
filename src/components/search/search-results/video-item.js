import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { VideoWrapper, VideoImage, InnerWrapper, Title, Description } from './video-item.styles'

class VideoItem extends PureComponent {
  showVideo = () => {
    const { selectVideo, video: { id } } = this.props
    selectVideo(id)
  }

  render() {
    const { title, thumbnail, description } = this.props.video
    return (
      <VideoWrapper onClick={this.showVideo}>
        <InnerWrapper>
          <VideoImage thumbnail={thumbnail} />
          <Title>{title}</Title>
          <Description>{description}</Description>
        </InnerWrapper>
      </VideoWrapper>
    )
  }
}

VideoItem.propTypes = {
  results: ImmutablePropTypes.contains(
    ImmutablePropTypes.recordOf({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      thumbnail: PropTypes.string.isRequired,
    })
  ),
}

export default VideoItem
