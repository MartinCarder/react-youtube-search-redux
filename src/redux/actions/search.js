import { youtubeSearch } from '../services/youtube'
export const SEARCH_REQESTED = 'SEARCH_REQESTED'
export const SEARCH_FAILED = 'SEARCH_FAILED'
export const SEARCH_SUCCESS = 'SEARCH_SUCCESS'

export const searchRequested = searchQuery => {
  return dispatch => {
    return youtubeSearch(searchQuery)
      .then(data => dispatch(searchSuccess(data)))
      .catch(err => {
        dispatch(searchFailed(err))
      })
  }
}

export const searchFailed = message => ({
  type: SEARCH_FAILED,
  message,
})

export const searchSuccess = data => ({
  type: SEARCH_SUCCESS,
  data,
})
