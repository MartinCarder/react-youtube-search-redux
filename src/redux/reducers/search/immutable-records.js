import { Record } from 'immutable'

export const SearchResultRecord = Record(
  {
    id: null,
    title: '',
    publishedAt: '',
    description: '',
    thumbnail: '',
  },
  'SearchResultRecord'
)

export const reviveSearchResultRecord = video =>
  new SearchResultRecord({
    id: video.id.videoId,
    thumbnail: video.snippet.thumbnails.high.url,
    ...video.snippet,
  })
