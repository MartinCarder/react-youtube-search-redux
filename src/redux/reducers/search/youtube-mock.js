export const youtubeMock = [
  {
    id: { kind: 'youtube#video', videoId: 'hY7m5jjJ9mM' },
    snippet: {
      channelId: 'UC9obdDRxQkmn_4YpcBMTYLw',
      channelTitle: 'Tiger FunnyWorks',
      description:
        'Cats are amazing creatures because they make us laugh all the time! Watching funny cats is the hardest try not to laugh challenge! Just look how all these cats & kittens play, fail, get along...',
      liveBroadcastContent: 'none',
      publishedAt: '2017-05-31T09:30:02.000Z',
      thumbnails: {
        high: {
          url: 'https://i.ytimg.com/vi/hY7m5jjJ9mM/hqdefault.jpg',
        },
      },
    },
  },
  {
    id: { kind: 'youtube#video', videoId: 'WEkSYw3o5is' },
    snippet: {
      channelId: 'UC9obdDRxQkmn_4YpcBMTYLw',
      channelTitle: 'Tiger FunnyWorks',
      description:
        'Well well well, cats for you again. But this time, even better, even funnier, even more hilarious! We know you like these furries the most, so thats why we keep making such compilations :)...',
      liveBroadcastContent: 'none',
      publishedAt: '2017-10-07T11:00:02.000Z',
      thumbnails: {
        high: {
          url: 'https://i.ytimg.com/vi/WEkSYw3o5is/hqdefault.jpg',
        },
      },
    },
  },
]
