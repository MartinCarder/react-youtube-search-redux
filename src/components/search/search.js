import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { searchRequested } from '../../redux/actions/search'
import SearchBar from './search-bar/search-bar'

export class Search extends PureComponent {
  render() {
    const { startSearch } = this.props
    return (
      <div>
        <SearchBar startSearch={startSearch} />
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => ({
  startSearch: searchTerm => dispatch(searchRequested(searchTerm)),
})
export default connect(null, mapDispatchToProps)(Search)
