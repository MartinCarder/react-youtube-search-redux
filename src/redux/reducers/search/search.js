import { Map, List } from 'immutable'
import { reviveSearchResultRecord } from './immutable-records'
import { SEARCH_REQESTED, SEARCH_FAILED, SEARCH_SUCCESS } from '../../actions/search'

const initialState = Map({
  searchTerm: '',
  status: null,
  data: List(),
  nextPageToken: null,
})

const searchReducer = (state = initialState, action) => {
  switch (action.type) {
    case SEARCH_REQESTED:
      return state.set('status', 'loading')
    case SEARCH_FAILED:
      return state.set('status', 'error')
    case SEARCH_SUCCESS:
      const { data } = action
      return state.withMutations(searchState => {
        searchState.set('data', List(data.items.map(video => reviveSearchResultRecord(video))))
        searchState.set('nextPageToken', data.nextPageToken)
        searchState.set('status', null)
      })
    default:
      return state
  }
}

export default searchReducer
