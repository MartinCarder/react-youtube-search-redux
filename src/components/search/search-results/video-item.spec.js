import React from 'react'
import { shallow, render } from 'enzyme'
import VideoItem from './video-item'
import { VideoWrapper } from './video-item.styles'
import { SearchResultRecord } from '../../../redux/reducers/search/immutable-records'

const video = new SearchResultRecord({
  id: '1234',
  title: 'Another cat video',
  description: 'Cats cats cats!!',
  publishedAt: '2017-05-31T09:30:02.000Z',
  thumbnail: 'image',
})

describe('<VideoItem>', () => {
  it('component renders correctly', () => {
    const wrapper = render(<VideoItem video={video} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('selectVideo called with correct id when video clicked', () => {
    const selectVideo = jest.fn()
    const wrapper = shallow(<VideoItem video={video} selectVideo={selectVideo} />)

    wrapper.find(VideoWrapper).simulate('click')
    expect(selectVideo.mock.calls[0][0]).toBe('1234')
  })
})
