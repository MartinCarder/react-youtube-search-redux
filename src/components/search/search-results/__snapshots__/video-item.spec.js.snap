// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoItem> component renders correctly 1`] = `
.c0 {
  -webkit-flex: 0 0 33.33%;
  -ms-flex: 0 0 33.33%;
  flex: 0 0 33.33%;
  box-sizing: border-box;
  padding: 7px 5px;
  width: 33.33%;
  overflow: hidden;
  position: relative;
}

.c2 {
  width: 100%;
  position: relative;
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
  background-image: url('image');
}

.c2:after {
  content: '';
  display: block;
  padding-top: 56.2%;
}

.c1 {
  padding: 2px;
  height: 100%;
  cursor: pointer;
}

.c1:before {
  content: '';
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: #ffffff;
  z-index: -1;
  margin: 7px 5px;
}

.c3 {
  font-weight: bold;
  font-size: 17px;
  text-align: center;
  color: #167ac6;
  margin: 15px;
}

.c4 {
  font-size: 13px;
  margin: 0 15px 15px 15px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      
    </div>
    <h5
      class="c3"
    >
      Another cat video
    </h5>
    <p
      class="c4"
    >
      Cats cats cats!!
    </p>
  </div>
</div>
`;
